### PR TITLE
Add resource profiles to control the resourceRequirements for ceph pods

### PR DESCRIPTION
**Pull request by malayparida2000**
```
• Add an storageCluster api field to specify the resource profile.
• The field can only take values "lean", "balanced" and "performance".
• Accordingly, the resourceRequirements for the ceph daemons are set,
  currently only mgr, mon, osd, mds and rgw pods are supported.
• It's checked if custom resourceRequirements are specified in the CR,
  if yes, then those are used. If not, then profiles are used.
• If no resource profile is specified, balanced profile values are used.
```

### CODE REVIEW

File: controllers/storagecluster/resourceprofile.go
Comment (travisn) [inline controllers/storagecluster/resourceprofile.go:28]: Why do we need to verify every detail of the resources applied? Can't we just confirm that the reconcile is completed (the CephCluster CR completion status), and perhaps as one extra step confirm that no ceph pods are pending? We should trust that the rook operator will do its job. A simple pod query with the labelSelector `rook_cluster=<namespace>` may be sufficient, then check no pods are in pending status.

Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go:28]: Tried this, Here are the issues

1. When the cephcluster spec is updated it takes a few seconds for rook to mark it as progressing, So in this brief moment both our checks pass i. Cephcluster being Ready, ii, None of the pods with the label are pending. Hence ocs operator thinks the resource profile has been applied successfully. While the truth is nothing has started out yet.

2. Even if we bypass the above problem somehow when rook starts processing the changes, The pod restarts are done one by one and in between two restarts there are brief moments when the cephcluster reports as ready & non of the pods with the label are pending. So OCS can again take it as a false positive & report the new profile has been applied already when it's actually in the halfway stage.

3. When we are trying to do a get on the pods with the label rook_cluster=<namespace> we are also getting crashcollectors, exporters, osd-prepare pods which are not actually being affected by this profile mechanism & if they are pending due to some other reason ocs will still report that it failed to apply the resource profile while it's not the case.

File: api/v1/storagecluster_types.go
Comment (malayparida2000) [reply api/v1/storagecluster_types.go:182]: I think marking something as omitempty makes it optional as well,
ref-https://book.kubebuilder.io/reference/markers.html#:~:text=The%20most%20reliable%20way%20in%201.x%20to%20get%20%2Boptional%20is%20omitempty.

**Comment by malayparida2000** [comment]
/hold
Remove hold after https://github.com/red-hat-storage/ocs-operator/pull/2257 is merged

File: controllers/storagecluster/resourceprofile.go
Comment (travisn) [reply controllers/storagecluster/resourceprofile.go:28]: Here is another approach... The CephCluster CR allows setting [labels](https://github.com/rook/rook/blob/master/deploy/examples/cluster.yaml#L202-L208) for different types of daemons. What if you set the name of the profile as a label for `all` components? For example, a label for `odf-profile=lean|balanced|performance`. Then all you need is a single query to see if daemons have the expected profile.

Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go:27]: Tried it, Using the labels looks like a better idea than comparing the ResourceRequirements. Have updated the code. But, instead of all have given labels to specifically mon, mgr, osd on the cephcluster CR & mds on the filesystem CR, rgw on the objectstore CR.

File: controllers/storagecluster/resourceprofile.go
Comment (iamniting) [inline controllers/storagecluster/resourceprofile.go:114]: Lets change this func name, maybe `getRunningPods`

File: controllers/storagecluster/resourceprofile.go
Comment (iamniting) [inline controllers/storagecluster/resourceprofile.go:22]: Can you define these as an error?

File: controllers/storagecluster/resourceprofile.go
Comment (iamniting) [inline controllers/storagecluster/resourceprofile.go:65]: Can you pls explain this regex via adding the comment in the code?

File: controllers/storagecluster/resourceprofile.go
Comment (iamniting) [inline controllers/storagecluster/resourceprofile.go:38]: I think we can create a common func for verifying these pods.

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/cephcluster.go
Comment (alfonsomthd) [reply controllers/storagecluster/cephcluster.go:469]: @iamniting from the UI perspective, when creating a new cluster the customer will always have to choose a profile and the backend will receive it.
For existing clusters that do not have a `resourceProfile` set yet, we should determine if they were deployed as `minimal` i.e. setting these `ResourceRequirements` (that no longer apply once this feature is merged):
https://github.com/red-hat-storage/odf-console/blob/master/packages/odf/types/common.ts#L42-L74

Comment (iamniting) [reply controllers/storagecluster/cephcluster.go:469]: @alfonsomthd We should not determine if they are in minimal mode or not as we are setting the values for 5 daemons. There are fewer chances that we will have the same values as the `minimal`. Moreover, if someone has set some values that are not default they should always be treated as custom instead of assigning any profile to them automatically.

If someone has not set any values we should treat it as balanced which is default now.

Comment (malayparida2000) [reply controllers/storagecluster/cephcluster.go]: Hey @alfonsomthd , The minimal mode that was being deployed by the UI was nothing but specifying the daemon resources manually in the storagecluster CR. With this profile mechanism also we respect the manual values added in the CR first ahead of the profiles. So if there are any clusters present that had deployed on that minimal mode from UI their clusters will work fine even after the upgrade as their daemons will still use the values that are present in the CR.

**Comment by malayparida2000** [comment]
@iamniting Along with addressing all your suggestions I have made the 4th commit that was about Ensuring the resource profile changes are applied correctly without any issues, a separate PR here https://github.com/red-hat-storage/ocs-operator/pull/2283. I felt the scope of this PR was getting too big & that was becoming a problem in getting it reviewed. Also as this is a blocker PR for the UI changes making it small with only the necessary things will ensure it getting merged sooner.
Once this PR is merged we can go ahead with the handling the profile change PR.

File: controllers/storagecluster/cephcluster.go
Comment (alfonsomthd) [reply controllers/storagecluster/cephcluster.go]: @iamniting What I want to point out is that if we deployed as minimal was due to the cluster having very low resources. If now we blindly apply Balanced to that cluster, chances are that we could face pod scheduling issues, right?

Comment (iamniting) [reply controllers/storagecluster/cephcluster.go]: We will not make it balanced. It will still stay as it is. As custom resources were set there.

**Review by iamniting — APPROVED** [review]

**Pull request merged**
